Libraries/MLXVLM/Models/Paligemma.swift

```diff
@@ -0,0 +1,743 @@
+
+        return prepareInputsForMultimodal(
+            imageFeatures: imageFeatures, inputEmbedding: inputEmbedding,
+            inputIds: inputIds, attentionMask: mask)
+    }
+
+    private func prepareInputsForMultimodal(
+        imageFeatures: MLXArray, inputEmbedding: MLXArray, inputIds: MLXArray,
+        attentionMask: MLXArray
+    ) -> (MLXArray, MLXArray) {
+        let embedDimension = imageFeatures.dim(2)
+        let (batchSize, sequenceLength) = inputIds.shape2
+        var scaledImageFeatures = imageFeatures / pow(Float(config.hiddenSize), 0.5)
+        var finalEmbedding = zeros([batchSize, sequenceLength, embedDimension])
+
+        let textMask = (inputIds .!= config.imageTokenIndex) & (inputIds .!= config.padTokenId)
+        let imageMask = inputIds .== config.imageTokenIndex
+        let padMask = inputIds .== config.padTokenId
+
+        // expand masks to match embedding dimension
+        var textMaskExpanded = expandedDimensions(textMask, axis: -1)
+        textMaskExpanded = repeated(textMaskExpanded, count: embedDimension, axis: -1)
+        var padMaskExpanded = expandedDimensions(padMask, axis: -1)
+        padMaskExpanded = repeated(padMaskExpanded, count: embedDimension, axis: -1)
+
+        // insert padding and text token embeddings
+        finalEmbedding = which(textMaskExpanded, inputEmbedding, finalEmbedding)
+        finalEmbedding = which(padMaskExpanded, zeros(like: finalEmbedding), finalEmbedding)
+
+        let padSize = finalEmbedding.dim(1) - scaledImageFeatures.dim(1)
+        scaledImageFeatures = padded(scaledImageFeatures, widths: [0, .init((0, padSize)), 0])
+
+        // insert image embeddings - the image mask is always less or equal to the sentence in length
+        var imageMaskExpanded = expandedDimensions(imageMask, axis: -1)
+        imageMaskExpanded = repeated(imageMaskExpanded, count: embedDimension, axis: -1)
+        finalEmbedding = which(imageMaskExpanded, scaledImageFeatures, finalEmbedding)
+
+        finalEmbedding = which(padMaskExpanded, zeros(like: finalEmbedding), finalEmbedding)
+
+        let attentionMaskExpanded1 = expandedDimensions(attentionMask, axis: 1)
+        let attentionMaskExpanded2 = expandedDimensions(attentionMask, axis: 2)
+        var finalAttentionMask4d = attentionMaskExpanded1 * attentionMaskExpanded2
+        finalAttentionMask4d = expandedDimensions(finalAttentionMask4d, axis: 1)
+
+        return (finalEmbedding, finalAttentionMask4d)
+    }
+
+    public func prepare(_ input: LMInput, cache: [any KVCache], windowSize: Int?) throws
+        -> PrepareResult
+    {
+        guard let image = input.image else { throw VLMError.imageRequired }
+        guard let mask = input.text.mask else { throw VLMError.maskRequired }
+        let inputIds = input.text.tokens
+
+        let (inputEmbedding, finalAttentionMask4d) = inputEmbeddings(
+            inputIds: inputIds, pixelValues: image.pixels, mask: mask)
```

Comment:
We might want to cast the `inputEmbedding` to the LM dtype as well (get it from the embedding layer weight or something).. just in case they have different types.